feature/blur

abdo-essam opened the pull request:
Refactor ImageViewFilter and introduce caching

- **ImageModerationController:** A new `ImageModerationController` class is introduced to encapsulate the image processing and moderation logic. This controller handles caching, loading states, and error handling.
- **ModerationCacheManager:** A `ModerationCacheManager` using `LruCache` is implemented to cache the results of image moderation. This helps to avoid redundant processing and improves performance for previously seen images.
- **ImageModerationState:** A data class `ImageModerationState` is defined to represent the different states of the image moderation process (loading, processing, moderated, error).
- **ModeratedImage Composable:** A new `ModeratedImage` composable is created to handle the display of the image based on its moderation state (original, blurred).
- **ImageViewFilter Simplification:** The `ImageViewFilter` composable is simplified to primarily delegate to the `ImageModerationController` and `ModeratedImage`. It now uses `LaunchedEffect` for asynchronous operations.
- **Error Handling and Loading States:** Improved error handling and clearer loading state management are implemented.
- **Code Structure:** The code is restructured for better organization and readability.

The `ActorItem` composable is updated to wrap `ImageViewFilter` in a `Box` to ensure proper layout.
The `TvShowDetails` screen is updated to use the new `loadingContent` and `errorContent` parameters of `ImageViewFilter`.

abdo-essam committed:
feat: Implement ImageHaramBlur feature

This commit introduces the ImageHaramBlur feature, which
provides image moderation capabilities. It includes face
detection, gender detection, and content detection to blur
images based on defined criteria.

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into develop

abdo-essam committed:
Merge remote-tracking branch 'origin/develop' into develop

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into develop

abdo-essam committed:
feat: Implement image blurring feature

This commit introduces an image blurring feature to the app.

The `ImageViewFilter` composable is added, which uses a
TensorFlow Lite model to detect inappropriate content in images
and blur it if necessary. This composable is now used in
`CategoryItem` and `HomeCard` to moderate displayed images.

Also, updated build configurations to include necessary
dependencies and configurations for the image blurring feature.

abdo-essam committed:
feat: Enhance image moderation by adding showImage state

abdo-essam committed:
fix: Improve image moderation stability and performance

This commit introduces several improvements to the image moderation feature:

- **Enhanced Stability:**
    - Addresses potential crashes by adding checks for drawable validity and bitmap recycling before processing.
    - Ensures the image is displayed even if an error occurs during moderation.
    - Implements safer cleanup logic for the `ImageModerationProcessor` by waiting for active moderation tasks to complete before releasing resources. This is managed using a coroutine Job and a Mutex.
- **Performance:**
    - Enables disk caching for `AsyncImage` to improve loading times for previously processed images.
- **General Improvements:**
    - Adds error logging in `ImageViewFilter` for better debugging.
    - Increases the minimum SDK version for the `ImageHaramBlur` feature to 26.
    - Adds necessary Google Play Services version and Firebase Performance metadata to the `AndroidManifest.xml`.

abdo-essam committed:
feat(ui): Conditionally enable crossfade

Crossfade is now disabled when moderation is enabled to
improve initial load performance. Also ensure that image
shows even on error.

abdo-essam committed:
feat(ui): Process image only if valid drawable

The code now checks for drawable validity (intrinsic width and
height) before processing to prevent exceptions. Also removed
unused Log import and Log.e call.

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
feat: Remove ImageHaramBlur module

Removes the ImageHaramBlur module from the AppConfig and
build.gradle.kts. This module is likely being deprecated or
moved.

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

Al-Taie committed:
Merge branch 'develop' into feature/blur

Al-Taie committed:
Merge branch 'develop' into feature/blur

abdo-essam committed:
feat(ui): Implement bitmap blurring for API < 31

Adds a `blurBitmap` utility function and integrates it into
`ImageViewFilter` to blur images on devices with API levels below 31,
as the Compose Modifier.blur is only available from API 31.
Also, increased the default blur strength.

abdo-essam committed:
Merge branch 'feature/blur' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
chore(build): Remove kotlinx-kover plugin

The kotlinx-kover plugin was removed from the build.gradle.kts
file. This plugin is not needed in the current configuration.

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
feat(ui): Implement blur effect in ImageViewFilter

Add blur functionality to the ImageViewFilter composable.
This allows applying a blur effect to images displayed
using this composable.

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
feat(ui): Improve image loading and moderation

Refactor ImageViewFilter to handle loading states and simplify
image display logic.  The changes ensure proper handling of
image loading, moderation, and blurring, especially for older
Android versions. Loading state is now also exposed via a callback.

Update HomeCard to use onLoadingStateChange.

abdo-essam committed:
feat(blur): Disable CoroutineCreationDuringComposition lint

Disable the `CoroutineCreationDuringComposition` lint check in
`ImageHaramBlur` module. This prevents false positives and allows
for intentional coroutine usage within composables in this feature.

abdo-essam committed:
feat(blur): Update coroutine dispatcher for blurBitmap and add error logging in ImageViewFilter

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

# Conflicts:
#	designSystem/src/main/java/com/london/designsystem/component/ActorItem.kt
#	feature/ImageHaramBlur/src/main/java/com/ae/imageharamblur/ui/ImageViewFilter.kt

karimdeabes committed:
fix: error image alignment center in dark and light

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

abdo-essam committed:
refactor(blur): Refactor ImageViewFilter and introduce caching

This commit refactors the `ImageViewFilter` composable and introduces several improvements:

- **ImageModerationController:** A new `ImageModerationController` class is introduced to encapsulate the image processing and moderation logic. This controller handles caching, loading states, and error handling.
- **ModerationCacheManager:** A `ModerationCacheManager` using `LruCache` is implemented to cache the results of image moderation. This helps to avoid redundant processing and improves performance for previously seen images.
- **ImageModerationState:** A data class `ImageModerationState` is defined to represent the different states of the image moderation process (loading, processing, moderated, error).
- **ModeratedImage Composable:** A new `ModeratedImage` composable is created to handle the display of the image based on its moderation state (original, blurred).
- **ImageViewFilter Simplification:** The `ImageViewFilter` composable is simplified to primarily delegate to the `ImageModerationController` and `ModeratedImage`. It now uses `LaunchedEffect` for asynchronous operations.
- **Error Handling and Loading States:** Improved error handling and clearer loading state management are implemented.
- **Code Structure:** The code is restructured for better organization and readability.

The `ActorItem` composable is updated to wrap `ImageViewFilter` in a `Box` to ensure proper layout.
The `TvShowDetails` screen is updated to use the new `loadingContent` and `errorContent` parameters of `ImageViewFilter`.
A comment in `designSystem/build.gradle.kts` is updated.

ahmeddsamii reviewed (dismissed):
Good job abdo 👏

Al-Taie committed:
Merge branch 'develop' into feature/blur

Al-Taie committed:
Merge branch 'develop' into feature/blur

AsmaaKhaledGomaa commented on designSystem/build.gradle.kts:
```diff
@@ -48,6 +48,6 @@ dependencies {
 
-    //lottie animation
+    // lottie animation
     implementation(libs.lottie.compose)
```
We are already using:
implementation(libs.bundles.ui.utils)

This bundle includes:
ui-utils = ["lottie-compose"]

Therefore, you don’t need to add this dependency separately. You can delete it.

abdo-essam replied:
I only added a space in the comment (changed //lottie to // lottie) for formatting consistency. 
 I didn't actually add the dependency as it was already there. 
 But good to know it's included in the ui-utils bundle - i removed this redundant line completely in the branch.    Appreciate the attention to detail Asmaa 👏

AsmaaKhaledGomaa commented on designSystem/src/main/java/com/london/designsystem/component/ActorItem.kt:
```diff
@@ -57,26 +59,28 @@ private fun ActorImage(imageRes: Any) {
+    Box {
+        ImageViewFilter(
+            model = imageRes,
+            contentDescription = "actorImage",
```
Kindly ensure that all contentDescription values are properly translated

abdo-essam replied:
Thanks for the suggestion, the content description is purely decorative (it doesn't convey meaningful information).

marwanqashwa reviewed (commented):
Good job just solve build problem

Abdulrahman-Ragab-01 reviewed (dismissed):
Great Job

abdo-essam committed:
Merge branch 'develop' of https://github.com/LondonSquad/Novix into feature/blur

# Conflicts:
#	presentation/src/main/java/com/london/presentation/screen/details/tvshow/tvshowdetails/TvShowDetails.kt

abdo-essam committed:
refactor(tvshowdetails): Clean up code by removing commented regions

abdo-essam committed:
feat(designSystem): Remove Lottie dependency

Removes the Lottie animation dependency from the designSystem
module. This dependency is no longer needed in the module.

Abdulrahman-Ragab-01 committed:
Merge branch 'develop' into feature/blur

Al-Taie committed:
Merge branch 'develop' into feature/blur

Al-Taie committed:
Merge branch 'develop' into feature/blur

Al-Taie committed:
Merge branch 'develop' into feature/blur

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

abdo-essam committed:
feat(lint): disable StateFlowValueCalledInComposition lint rule
